File: ds/hash_serialize.go
package ds

import (
	"bufio"
	"fmt"
	"os"
)

type HashSerializer struct{}

func NewHashSerializer() *HashSerializer {
	return &HashSerializer{}
}

func (s *HashSerializer) SaveToFile(hash interface{}, filename string) error {
	switch h := hash.(type) {
	case *ChainHash:
		return s.saveChainHash(h, filename)
	case *OpenHash:
		return s.saveOpenHash(h, filename)
	default:
		return fmt.Errorf("unsupported hash type")
	}
}

func (s *HashSerializer) LoadFromFile(hash interface{}, filename string) error {
	switch h := hash.(type) {
	case *ChainHash:
		return s.loadChainHash(h, filename)
	case *OpenHash:
		return s.loadOpenHash(h, filename)
	default:
		return fmt.Errorf("unsupported hash type")
	}
}

func (s *HashSerializer) SaveToBinaryFile(hash interface{}, filename string) error {
	switch h := hash.(type) {
	case *ChainHash:
		return s.saveChainHashBinary(h, filename)
	case *OpenHash:
		return s.saveOpenHashBinary(h, filename)
	default:
		return fmt.Errorf("unsupported hash type")
	}
}

func (s *HashSerializer) LoadFromBinaryFile(hash interface{}, filename string) error {
	switch h := hash.(type) {
	case *ChainHash:
		return s.loadChainHashBinary(h, filename)
	case *OpenHash:
		return s.loadOpenHashBinary(h, filename)
	default:
		return fmt.Errorf("unsupported hash type")
	}
}

func (s *HashSerializer) saveChainHash(ch *ChainHash, filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	fmt.Fprintln(file, ch.Cap)
	count := 0
	for _, n := range ch.Table {
		for cur := n; cur != nil; cur = cur.Next {
			count++
		}
	}
	fmt.Fprintln(file, count)

	for _, n := range ch.Table {
		for cur := n; cur != nil; cur = cur.Next {
			WriteStringText(file, cur.Key)
			WriteStringText(file, cur.Val)
		}
	}
	return nil
}

func (s *HashSerializer) loadChainHash(ch *ChainHash, filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	sc := bufio.NewScanner(file)
	var capVal, count int
	if sc.Scan() {
		fmt.Sscanf(sc.Text(), "%d", &capVal)
	}

	ch.Cap = capVal
	ch.Table = make([]*ChainNode, ch.Cap)

	if sc.Scan() {
		fmt.Sscanf(sc.Text(), "%d", &count)
	}

	for i := 0; i < count; i++ {
		k := ReadStringText(sc)
		v := ReadStringText(sc)
		ch.Insert(k, v)
	}
	return nil
}

func (s *HashSerializer) saveChainHashBinary(ch *ChainHash, filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	WriteInt32(file, int32(ch.Cap))
	count := 0
	for _, n := range ch.Table {
		for cur := n; cur != nil; cur = cur.Next {
			count++
		}
	}
	WriteInt32(file, int32(count))

	for _, n := range ch.Table {
		for cur := n; cur != nil; cur = cur.Next {
			WriteString(file, cur.Key)
			WriteString(file, cur.Val)
		}
	}
	return nil
}

func (s *HashSerializer) loadChainHashBinary(ch *ChainHash, filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	capVal, _ := ReadInt32(file)
	ch.Cap = int(capVal)
	ch.Table = make([]*ChainNode, ch.Cap)

	count, _ := ReadInt32(file)
	for i := 0; i < int(count); i++ {
		k, _ := ReadString(file)
		v, _ := ReadString(file)
		ch.Insert(k, v)
	}
	return nil
}

func (s *HashSerializer) saveOpenHash(oh *OpenHash, filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	fmt.Fprintln(file, oh.Cap)
	count := 0
	for _, e := range oh.Table {
		if e.Used {
			count++
		}
	}
	fmt.Fprintln(file, count)

	for _, e := range oh.Table {
		if e.Used {
			WriteStringText(file, e.Key)
			WriteStringText(file, e.Val)
		}
	}
	return nil
}

func (s *HashSerializer) loadOpenHash(oh *OpenHash, filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	sc := bufio.NewScanner(file)
	var capVal, count int
	if sc.Scan() {
		fmt.Sscanf(sc.Text(), "%d", &capVal)
	}
	oh.Cap = capVal
	oh.Table = make([]OpenEntry, capVal)

	if sc.Scan() {
		fmt.Sscanf(sc.Text(), "%d", &count)
	}

	for i := 0; i < count; i++ {
		k := ReadStringText(sc)
		v := ReadStringText(sc)
		oh.Insert(k, v)
	}
	return nil
}

func (s *HashSerializer) saveOpenHashBinary(oh *OpenHash, filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	WriteInt32(file, int32(oh.Cap))
	count := 0
	for _, e := range oh.Table {
		if e.Used {
			count++
		}
	}
	WriteInt32(file, int32(count))

	for _, e := range oh.Table {
		if e.Used {
			WriteString(file, e.Key)
			WriteString(file, e.Val)
		}
	}
	return nil
}

func (s *HashSerializer) loadOpenHashBinary(oh *OpenHash, filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	capVal, _ := ReadInt32(file)
	oh.Cap = int(capVal)
	oh.Table = make([]OpenEntry, oh.Cap)

	count, _ := ReadInt32(file)
	for i := 0; i < int(count); i++ {
		k, _ := ReadString(file)
		v, _ := ReadString(file)
		oh.Insert(k, v)
	}
	return nil
}

File: hash_serialize.cpp
#include "hash_serialize.h"
#include "serialize.h"
#include <iostream>

using namespace std;

// Сохранение ChainHash в текстовый файл
void HashSerializer::saveToFile(const ChainHash& table, const string& filename) {
    ofstream file(filename);
    if (!file) return;

    file << table.cap << "\n";

    int count = 0;
    for (int i = 0; i < table.cap; ++i) {
        ChainHash::Node* curr = table.table[i];
        while (curr) {
            count++;
            curr = curr->next;
        }
    }
    file << count << "\n";
    
    for (int i = 0; i < table.cap; ++i) {
        ChainHash::Node* curr = table.table[i];
        while (curr) {
            writeStringText(file, curr->key);
            writeStringText(file, curr->val);
            curr = curr->next;
        }
    }
}

// Загрузка ChainHash из текстового файла
void HashSerializer::loadFromFile(ChainHash& table, const string& filename) {
    ifstream file(filename);
    if (!file) return;

    int newCap;
    file >> newCap;
    string dummy; getline(file, dummy);
    
    table.clearAndResize(newCap);
    
    int count;
    file >> count;
    getline(file, dummy);
    
    for (int i = 0; i < count; ++i) {
        string key = readStringText(file);
        string val = readStringText(file);
        if (file.fail()) break;
        table.insert(key, val);
    }
}

// Бинарное сохранение ChainHash
void HashSerializer::saveToBinaryFile(const ChainHash& table, const string& filename) {
    ofstream file(filename, ios::binary | ios::trunc);
    if (!file) return;
    
    file.write(reinterpret_cast<const char*>(&table.cap), sizeof(table.cap));
    
    int count = 0;
    for (int i = 0; i < table.cap; ++i) {
        ChainHash::Node* curr = table.table[i];
        while (curr) {
            count++;
            curr = curr->next;
        }
    }
    file.write(reinterpret_cast<const char*>(&count), sizeof(count));
    
    for (int i = 0; i < table.cap; ++i) {
        ChainHash::Node* curr = table.table[i];
        while (curr) {
            writeString(file, curr->key);
            writeString(file, curr->val);
            curr = curr->next;
        }
    }
}

// Бинарная загрузка ChainHash
void HashSerializer::loadFromBinaryFile(ChainHash& table, const string& filename) {
    ifstream file(filename, ios::binary);
    if (!file) return;
    
    int newCap;
    file.read(reinterpret_cast<char*>(&newCap), sizeof(newCap));
    table.clearAndResize(newCap);
    
    int count;
    file.read(reinterpret_cast<char*>(&count), sizeof(count));
    
    if (file.fail()) return;
    for (int i = 0; i < count; ++i) {
        string key = readString(file);
        string val = readString(file);
        if (file.fail()) break;
        table.insert(key, val);
    }
}

// Сохранение OpenHash в текстовый файл
void HashSerializer::saveToFile(const OpenHash& table, const string& filename) {
    ofstream file(filename);
    if (!file) return;
    file << table.cap << "\n";
    
    int count = 0;
    for (int i = 0; i < table.cap; ++i) {
        if (table.table[i].used) count++;
    }
    file << count << "\n";

    for (int i = 0; i < table.cap; ++i) {
        if (table.table[i].used) {
            writeStringText(file, table.table[i].key);
            writeStringText(file, table.table[i].val);
        }
    }
}

// Загрузка OpenHash из текстового файла
void HashSerializer::loadFromFile(OpenHash& table, const string& filename) {
    ifstream file(filename);
    if (!file) return;
    
    int newCap;
    file >> newCap;
    string dummy; getline(file, dummy);
    
    table.clearAndResize(newCap);
    
    int count;
    file >> count;
    getline(file, dummy);
    
    if (file.fail()) return;
    for (int i = 0; i < count; ++i) {
        string key = readStringText(file);
        string val = readStringText(file);
        if (file.fail()) break;
        table.insert(key, val);
    }
}

// Бинарное сохранение OpenHash
void HashSerializer::saveToBinaryFile(const OpenHash& table, const string& filename) {
    ofstream file(filename, ios::binary | ios::trunc);
    if (!file) return;
    
    file.write(reinterpret_cast<const char*>(&table.cap), sizeof(table.cap));
    
    int count = 0;
    for (int i = 0; i < table.cap; ++i) {
        if (table.table[i].used) count++;
    }
    file.write(reinterpret_cast<const char*>(&count), sizeof(count));
    
    for (int i = 0; i < table.cap; ++i) {
        if (table.table[i].used) {
            writeString(file, table.table[i].key);
            writeString(file, table.table[i].val);
        }
    }
}

// Бинарная загрузка OpenHash
void HashSerializer::loadFromBinaryFile(OpenHash& table, const string& filename) {
    ifstream file(filename, ios::binary);
    if (!file) return;
    
    int newCap;
    file.read(reinterpret_cast<char*>(&newCap), sizeof(newCap));
    table.clearAndResize(newCap);
    
    int count;
    file.read(reinterpret_cast<char*>(&count), sizeof(count));
    
    for (int i = 0; i < count; ++i) {
        string key = readString(file);
        string val = readString(file);
        if (file.fail()) break;
        table.insert(key, val);
    }
}